### PR TITLE
[CR] prompt/query hotkey fix

### LINKED ISSUE #20065: Prompt/query hotkeys (e.g. when repairing) are "masked" by global binds (regression)
PR #20041 made some keys (most importantly, `2` and `q`, but also others) unavailable in queries, e.g. when repairing an item with a tool.

[Savefile](https://github.com/CleverRaven/Cataclysm-DDA/files/721553/test-repair-hotkeys.zip) demonstrates this. Character has a soldering iron and a pocket knife. Activating the iron and choosing the knife for repair brings up a repair/practice window, where `2` and `q` are shown as hotkeys for "repair as long as you can" and "cancel" respectively. Pressing those keys results in the selection moving down and no-action respectively.

This issue is only partially addressed by follow-up PR #20072 ATM - `q` does quit, but `2`/`8`, for example, are still mapped to scrolling.

-----

This issue has been split for easier tracking. Original many-edit braindump after the double break.

-----
-----

Several issues conflated.

System: Arch Linux, curses, self-built, current git master `0.C-20721-g82a1273ad9`.

### Cheat menu actions skip mappings

**See #20091.**

Works as intended? Missing keys are `2`/`8` (up/down on numpad), `f` (find), `j`/`k` (vi up/down), and `q` (quit).

Guess PR #20041, commit 18d78a0 ("Prevent hotkeys in uimenu that are already mapped to actions.") - confirmed by bisect.

Can provide save if necessary, but easier to test through cheat menu (that's what I did). Previously, all actions in it were mapped to keys (seemingly sequentially). Now, several actions don't have a key associated.

Commit can be easily reverted (will submit PR), but perhaps it was needed for fixing the bug?..

### Mapping `f` to "find" breaks cheat-wishing contained liquid

Incorrect! **See #20147.**

Now, both `/` and `f` are mapped to "find".

Could be ~~easily~~ remedied by mapping something else to "contained" (`c`?). Won't work well with wishing for a friendly monster, though.

### Crafting prompts don't map `q`

**See top of this report.**

Repairing an item with a soldering iron no longer maps "quit" to `q` key. The `q` is highlighted in the "Repairing/Practicing" prompt, but pressing the key does nothing.

For testing save, see below.

Pressing `ESC` does nothing either (pre-existing).

-----

Ping @BevapDin.

EDIT: Removed EDIT notes for readability, re-structured.

### LINKED ISSUE #20147: No special input in debug/cheat sub-menus (regression)
Fallout from PR #20041.

Certain sub-menus of the cheat menu, e.g. "Wish for an item" or "Spawn monster", rely on special key bindings, such as `f` for "contained item" or "friendly monster", `i`/`d` for group size change, etc.

Seems that key presses are no longer passed to the `key()` method of the relevant `*_callback` classes. The "global context" move up/down, ENTER, etc. do work.

Originally mentioned in #20065.

### PR DESCRIPTION
**Pull request by kolsurma**
Closes #20147, #20175 
#20065 is two issues in one, this fixes one, but not the other.

Should make #20251 full reversion of the new input system unnecessary. 

It does introduce 1 issue with the action menu which I'm trying to work around with the minimum changes required. When a character uses the action menu the following popup is always displayed, regardless of if it's required or not.
```
class actmenu_cb : public uimenu_callback

bool key( const input_event &event, int /*num*/, uimenu * /*menu*/ ) override {
            const std::string action = ctxt.input_to_action( event );
            if( action == "HELP_KEYBINDINGS" ) {
                ctxt.display_help();
                return true;
            }
            // Don't write a message if unknown command was sent
            // Only when an inexistent tool was selected
            auto itemless_action = am.find( action );
            if( itemless_action != am.end() ) {
                popup( _( "You do not have an item that can perform this action." ) );
            }
            return false;
        }
```

There's some weirdness with the way it handles non-enabled hotkeys (tool missing), but that's inherent in the way `uimenu::query` handles the selection of items.

### CODE REVIEW

**Comment by kevingranade** [comment]
Does it still mask keys?  E.g. debug->spawn item->f triggering search
instead of toggling containers?
Does it still introduce weird gaps in menu numbering?

In short, unless this fixes all the associated issues, I don't want to land
it because it will just make the solution in #20251 more difficult to apply.

**Comment by kolsurma** [comment]
>Does it still mask keys?  E.g. debug->spawn item->f triggering search
instead of toggling containers?

Nope, fixed that issue. Debugging the martial arts menu's '?' key now.

>Does it still introduce weird gaps in menu numbering?

Yep, 2 is unusable with the current scheme. We could just start counting with 3, or switch to abcd style. 

>In short, unless this fixes all the associated issues, I don't want to land
it because it will just make the solution in #20251 more difficult to apply.

Good point, if possible I'd rather have the @BevapDin's UI changes in a feature branch, but that would get stale with how quick changes are made on the master. I haven't seen what the code used to look like before the new UI, so I'm unsure how big a jump it was from that system to this one.

I tend to like to salvage working code when possible, but we shouldn't let the technical debt get too high here.

**Comment by kolsurma** [comment]
I think with these changes and #20174 that should get most of the issues from the new input system ironed out.

**Comment by keyspace** [comment]
I'll try a local merge of this PR and PR #20174 together on top of current master, will report back to confirm.

**Comment by keyspace** [comment]
ce2f17364fa175eb13f101b4f910f691e5c61350 - current `git master`, used as base.

First, merged `kolsurma/potential_keymap_fix`. Then merged `BevapDin/abc` on top.

-----

* Fixes #20065 (mostly)

`q` quits, as it should. `2` and `8` no longer scroll. When repairing an item with a soldering iron, `2` does execute "repair as long as possible", as it is mapped to.

Pressing any key that's not displayed in the query - like `8`, `b`, `m`, `\`` - seems to execute the currently selected action. So if it's been manually moved with arrow keys to "repeat until success/failure/level up", that will be executed.

`f`, `/`, `ESC` are exceptions, haven't looked at code if there are more. The latter doesn't do anything. The first two display a search input field which allows filtering out options.

-----

* Fixes #20091

Debug menu no longer skips mappings. Once again, they are all displayed sequentially (`1-90a-y`), as they used to, and pressing them executes the appropriate action.

`hjkl` (vi keys) no longer scroll, `2` and `8` also don't scroll.

-----

* Fixes #20147 (mostly)

`f` correctly toggles containers in "wish for item", or toggles friendliness in "spawn monster". Most other special mappings in "spawn monster" also work as expected.

In both menus, pressing `q` quits the menu _and_ executes the spawning action - which may actually be an "unintended feature", since `ESC` is available to "just quit".

-----

* Fixes #20175

Special techniques are once again properly displayed in martial art selection query/window/thing. Pressing `?` on a style with techniques displays them, as expected. Pressing `?` on a selection option without techniques performs no action.

**Comment by keyspace** [comment]
Not sure on the status of PR #20072 if this one and PR #20174 get merged. Possibly obsolete?

-----

> It does introduce 1 issue

Can reproduce this issue described in OP. It's with the `%` "item action menu", not the `a` "use/apply item" menu.

The message is displayed if the action-to-perform is selected via hotkey. Selecting it with up/down and pressing `Enter` works fine.

Will take a look.

**Comment by kolsurma** [comment]
@keyspace I'm a git nub, I've changed @BevapDin abc branch, should I merge it into this one or set a separate pull?

The 'q' key needs to be a reserved character for the UI menus, and it's been assigned to quit menus correctly now.

**Comment by keyspace** [comment]
I'm not sure what you're trying to achieve, honestly. :)

If you want commits from `BevapDin/abc` to be part of this PR, you could do something like (in a shell):

``` sh
git checkout potential_keymap_fix
git remote add bevapdin git@github.com:bevapdin/Cataclysm-DDA.git
git fetch bevapdin abc
git merge bevapdin/abc
git push -vn
git push -v
```

Alternatively, you could create a PR in your own repo, requesting (yourself) to merge that branch into your branch, then accept this web-side github PR in your repo.

Still, was this what you were asking about?..

**Comment by keyspace** [comment]
@kolsurma oh, you probably meant you've changed your locally checked-out copy of `BevapDin/abc`, and were asking whether you should open a separate superceding PR for that alone, or include it in this PR.

I'm not sure if there's a policy for that.

If "fixing it all in one PR" requires changes from `BevapDin/abc` and then some on top, then it sounds right to me for all the changes to be included in this PR. Then one PR will need merging, and one manual closing due to obsoletion, instead of two merging and one closing.

(BTW, in such a scenario, if the "composite" branch was merged first, and merging PR #20174 was attempted - it should have no effect, since the commits would have already been included. That is, of course, if they were included verbatim, otherwise there might be a conflict.)

**Comment by keyspace** [comment]
On issue outlined in OP - AFAIU that `key()` function's return value is checked by `uimenu::query()`, which expects a return value of `true` if the keys should be skipped. 

`am` (action map, `std::map< item_action_id, item_action >`) for `%` menu has 183 actions registered.

That's so high since it includes both items with registered actions, and all the "generic actions with bound keys", as constructed in in `game::item_action_menu()` (towards the end of `item_action,cpp`, same file as code in question). I don't know why it's done like this. The logic of that particular function might need review to address this bug.

-----

If the conditional is reversed as follows:

``` c++
68              if( itemless_action != am.end() ) {
69                  return false;
70              } else {
71                  popup( _( "You do not have an item that can perform this action." ) );
72                  return true;
73      	}
```

Then pressing `ESC` will display the message, and only choosing a valid action would exit the menu - which is undesirable.

-----

It may also be prudent to exit early from `game::item_action_menu()` if no items with actions are available, like this:

``` c++
234     add_msg( "%i in iactions", iactions.size() );
235     if( iactions.empty() ) {
236         popup( _( "You don't have any items with registered uses." ) );
237         return; // exit early
238     }
```

Again, I'm not sure if there being no `return` here is accidental, though.

**Comment by kolsurma** [comment]
I'm at the limit of my limited abilities. I'm unsure how to proceed on the issue that this creates. This should be moved to a different branch until the action menu issue is fixed instead of merged into master.

On the note of the action menu: the keymappings are horrifically un-intuitive. There are 9 different "Activate" mappings, 7 different "drink", etc.

**Comment by kolsurma** [comment]
The fix is an ugly series of checks, but it does seem to allow all menus to work correctly. I know it can be done better, but I reached the end of my abilities.

**Comment by keyspace** [comment]
@kolsurma don't worry, it's not _that_ ugly. The important part has been done - you've found the origin of the bug.

Got a bit of time now, so checking locally and documenting inline as to what's actually happening. Hopefully, will get to that "done better".

In the meanwhile, the following could be added to the preamble, or a link can be added to this comment instead, to save the hassle of formatting.

-----

In current master, generic `uimenu::query()` handles most input events first, and then passes the event to the callback if it can't handle it.

In this PR, the logic is reversed. The callback gets the event first, the only condition being that the callback is set at all. This allows cerain menus (mostly the ones that exhibited various input handling oddities) to use their callback for overriding "default" handling by `uimenu::query()`.

The callback returninig a boolean `false` signifies that the callback can't "handle the event completely". This is unchanged before or after the PR.

Currently there are these callbacks for menus:

``` sh
% grep -rE 'public uimenu_callback$' src
```
``` c++
src/ui.h:class pointmenu_cb : public uimenu_callback
src/player.cpp:class ma_style_callback : public uimenu_callback
src/item_action.cpp:class actmenu_cb : public uimenu_callback
src/wish.cpp:class wish_mutate_callback: public uimenu_callback
src/wish.cpp:class wish_monster_callback: public uimenu_callback
src/wish.cpp:class wish_item_callback: public uimenu_callback
```

**Comment by keyspace** [comment]
I can lo longer confirm that the current iteration of this PR fixes #20065. Either it's something on my end, or some breakage was introduced.

**Comment by kolsurma** [comment]
>I can lo longer confirm that the current iteration of this PR fixes #20065. Either it's something on my end, or some breakage was introduced.

Which part of #20065 isn't working for you? There's a few different issues in that one.

**Comment by keyspace** [comment]
> Which part of #20065 isn't working for you?

`q` no longer working as a hotkey/shortcut for quitting in the "repeat" menu. Which seems strange - this menu does not set a callback for its `uimenu` instance, [instead doing `addentry` manually](https://github.com/CleverRaven/Cataclysm-DDA/blob/234529d/src/activity_handlers.cpp#L1530).

(When mentioning #20065, I mean the part before the double break (first three paragraphs) - either of the two issues that are "local" to the repeat menu.)

**Comment by kolsurma** [comment]
I can't reproduce the error on my end. Ah, when you added `BevapDin\abc` did you also add my later commit 4e8633e? That adds `q` as an option to quit the menus along with `esc`.

**Comment by kevingranade** [comment]
I'm also getting the failure to quit from the solder->repair->repeat menu qith either q or Esc.

I'm also getting gaps in automatic ordering of hotkeys in the cut up menu.

If we have a systematic way to address these, I'm good to merge what we have and then chip away at the rest, I think this gives us that, but I'm not totally sure yet.

Thanks for keeping on this, I also don't want to revert that large swath of input overhaul.

**Comment by kevingranade** [comment]
Ok, found the problem with the repeat menu at least, adding 'q' as a mapping for QUIT has an unintended side effect of breaking q in menus where it's not configured to exit on unknown characters.

**Comment by kolsurma** [comment]
>I'm also getting gaps in automatic ordering of hotkeys in the cut up menu.

b,B,d,D,f,z,q are static:
**b**: butcher all entities
**B**: butcher first available entity
**d**: disassemble everything, repeating
**D**: disassemble everything, once
**f**: find (not displayed in menu)
**z**: cut up everything
**q**: cancel

Both b and d could be moved to other hotkeys, I feel that could be disruptive to existing players however.
The 'find' functionality can be very useful, although it's poorly documented in the UIs themselves and it tends to pop up in odd places.

**Comment by kevingranade** [comment]
This isn't the right change to repeat_menu, it only works because the 'q' in the local menu hapens to be the same as the 'q' in the global keybind, that "press q to cancel" is still never actually happening, and it makes that code fragile.  Was there a prticular bug fixed by adding 'q' as a global keybind?  If not, I'd rather just remove that global keybind and allow the local q mapping to work.

**Comment by kolsurma** [comment]
TL;dr: 
pre-uimenu update, `q`/`esc` cancel prompts. 

latest BevapDin uimenu update: `esc` cancels prompts, `q` does nothing unless explicitly added to list

this pr: `q`/`esc` cancel prompts. Bug roundup. Fragile code. Sorry for the wall-o'-text.


>This isn't the right change to repeat_menu, it only works because the 'q' in the local menu hapens to be the same as the 'q' in the global keybind, that "press q to cancel" is still never actually happening, and it makes that code fragile
>Was there a prticular bug fixed by adding 'q' as a global keybind? If not, I'd rather just remove that global keybind and allow the local q mapping to work.

I'm not opposed to dropping `q` from the global bind, it was just an expedient way to patch the q/esc => cancel issue in #20065 et al. To rehash the issue, `q` and `esc` in most menus allow the player to exit the menu, but not all menus have explicit `q  cancel` option. The `repeat_menu` does, which introduced an issue with `q`'s global bind. 2a4d478 brought the repeat menu inline with other menus that include a 'cancel', such as 
```c++
void game::butcher()
{
. . .
    if( corpses.size() > 1 || !disassembles.empty() || !salvageables.empty() ) {
        uimenu kmenu;
        . . .
        kmenu.addentry( CANCEL, true, 'q', _("Cancel"));
        kmenu.return_invalid = true;
        kmenu.query();
```

This is a pre-existing issue with `uimenu::query()`. The obvious choice is just to `addentry( CANCEL, true, 'q', _("Cancel"));` to every prompt, and may be where @BevapDin was heading with it. As you state, in it's current form this is fragile as it's never actually used when `q` is set as an exit hotkey and `return_invalid = true`.

I suppose the optimal path would be to go ahead and add the cancel entry into all menus.
